Rework RGBD sensor.

michalpelka opened the pull request:
The RGBD sensor is running only one pipeline with this change.

Plan :
 - [x] merge #103 

Signed-off-by: Michał Pełka <michal.pelka@robotec.ai>

michalpelka commented:
Unfortunately, currently code crashes with stack:
```__memmove_evex_unaligned_erms 0x00007ffff6e7b28c
memcpy string_fortified.h:29
AZ::RPI::AttachmentReadback::CopyBufferData AttachmentReadback.cpp:560
$_10::operator()() const AttachmentReadback.cpp:331
AZStd::Internal::INVOKE<…>($_10 &) invoke_traits.h:208
AZStd::invoke<…>($_10 &) invoke.h:54
AZStd::Internal::function_util::invoke_void_return_wrapper::call<…>($_10 &) function_template.h:41
AZStd::Internal::function_util::get_invoker::call<…>(AZStd::Internal::function_util::function_buffer &) function_template.h:172
AZStd::function_intermediate::operator() function_template.h:604
AZStd::function::operator()() const function_template.h:684
$_5::operator()() const Fence.cpp:122
AZStd::Internal::INVOKE<…>($_5 &&) invoke_traits.h:208
AZStd::invoke<…>($_5 &&) invoke.h:54
AZStd::Internal::apply_impl<…>($_5 &&, std::tuple<…> &&, std::integer_sequence<…>) tuple.h:281
AZStd::apply<…>($_5 &&, std::tuple<…> &&) tuple.h:289
<lambda#1>::operator()() thread_UnixLike.h:37
AZStd::Internal::thread_info_impl::execute() thread.h:193
AZStd::Internal::thread_run_function thread_UnixLike.cpp:35
start_thread 0x00007ffff6d60b43
clone3 0x00007ffff6df2a00```

michalpelka commented:
Problem with crash was solved by using CapturePassAttachmentWithCallback

VickyAtAZ commented on Gems/ROS2/Code/Source/Camera/CameraSensor.cpp:
```diff
@@ -240,4 +253,51 @@ namespace ROS2
         return "Color";
     };
 
+    CameraRGBDSensor::CameraRGBDSensor(const CameraSensorDescription& cameraSensorDescription)
```
Do you need to implement the GetPipelineTypeName() function to return a name different than "Color"?

michalpelka replied:
Thanks, It seems to be unnecessary, GetPipelineTypeName is used to create only pipeline instance names. I would not change it in this PR, since it is currently in codebase.

VickyAtAZ commented on Gems/ROS2/Code/Source/Camera/CameraSensor.cpp:
```diff
@@ -240,4 +253,51 @@ namespace ROS2
+                    message.width = descriptor.m_size.m_width;
+                    message.height = descriptor.m_size.m_height;
+                    message.step = message.width * Internal::BitDepth.at(format);
+                    message.data = std::vector<uint8_t>(result.m_dataBuffer->data(), result.m_dataBuffer->data() + result.m_dataBuffer->size());
```
This should be able to be simplified to 
`message.data = *result.m_dataBuffer;`

michalpelka replied:
We move data from AZStd::vector to std::vector. The simplification is, sadly, not feasible.

VickyAtAZ commented on Gems/ROS2/Code/Source/Camera/CameraSensor.h:
```diff
@@ -87,6 +88,27 @@ namespace ROS2
+        //! @param cameraPose - current camera pose from which the rendering should take place
+        //! @param callback - callback function object that will be called when capture is ready
+        //!                   it's argument is readback structure containing, among other thins, captured image
+        //! @param passName - pass name is pipeline, eg `DepthToLinearDepthPass`
```
nit: 'is' -> 'in'

VickyAtAZ commented on Gems/ROS2/Code/Source/Camera/CameraSensor.h:
```diff
@@ -113,4 +135,20 @@ namespace ROS2
+
+    private:
+        void ReadBackDepth(AZStd::function<void(const AZ::RPI::AttachmentReadback::ReadbackResult& result)> callback);
+        AZStd::shared_ptr<AZ::RPI::AttachmentReadback> m_readbackDepth;
```
This can be removed.

michalpelka replied:
Thanks! Removed

VickyAtAZ commented on Gems/ROS2/Code/Source/Camera/CameraSensor.cpp:
```diff
@@ -194,20 +194,33 @@ namespace ROS2
+                    message.width = descriptor.m_size.m_width;
+                    message.height = descriptor.m_size.m_height;
+                    message.step = message.width * Internal::BitDepth.at(format);
+                    message.data = std::vector<uint8_t>(result.m_dataBuffer->data(), result.m_dataBuffer->data() + result.m_dataBuffer->size());
```
This should be able to be simplified to
message.data = *result.m_dataBuffer;

michalpelka replied:
Unfortunately, cannot do that, `message.data` is std::vector (Provided by ROS2 RLCPP) and result.m_data buffer AZStd::vector. Your solution would be better but is not feasible.

VickyAtAZ reviewed (commented):
I can't find the change to ensure the camera render pipeline's name is unique. 
Did you change the implementation of GetFrameID some where to ensure the frameId is different for each camera?

michalpelka commented:
In the debugger, I can see that switching from Editor to Game mode I have dealock on `m_semaphore.acquire();`. Call stack is available [here](https://gist.github.com/michalpelka/2a9d5e35e8da6de120e09390b6d138fc).

michalpelka commented:
Good news, the ROS2 Camera component works stable against [o3de commit](https://github.com/o3de/o3de/commit/17e4615b4693141515908c7ea584fa8bb41054ce) . Thanks, @VickyAtAZ  for reviewing, I addressed your comments and suggestion.
I, unfortunately, found an issue with compatibility - the ROS2 Camera renders a pitch-black RGB image (depth is fine) when the main camera has Exposure adaptation enabled. I think there is some problem in custom pass templates in the resources of this Gem. You can reproduce easily with the Main of RosconDemo.
Rebased also on top of #103.

hultonha commented on Gems/ROS2/Code/Source/Camera/ROS2CameraSensorEditorComponent.h:
```diff
@@ -0,0 +1,58 @@
+        void Activate() override;
+        void Deactivate() override;
+
+        //! AzToolsFramework::Components::EditorComponentBase override
```
nit: For these override comments we just use `//` as opposed to `//!` for doxygen comments

michalpelka replied:
adjusted to review, thanks.

hultonha reviewed (approved):
These changes look good to me, would be great to hear from @VickyAtAZ if they're happy with the most recent changes (also might be worth taking a look at https://github.com/o3de/o3de/pull/14444 in case this is related to and kind of screen capture stuff). Thanks!

hultonha commented on Gems/ROS2/Code/Source/Camera/CameraSensor.cpp:
```diff
@@ -116,11 +117,12 @@ namespace ROS2
         m_scene = AZ::RPI::RPISystemInterface::Get()->GetSceneByName(AZ::Name("Main"));
 
-        const AZStd::string pipelineName = m_cameraSensorDescription.m_cameraName + "Pipeline" + GetPipelineTypeName();
+        m_pipelineName = m_cameraSensorDescription.m_cameraName + "Pipeline" + GetPipelineTypeName() +
```
Thank you! 🙂

VickyAtAZ commented:
@michalpelka The AR is failed with XR gem. I believe there is a commit which fixes it. You will need merge the change from development branch.

hultonha commented:
> @michalpelka The AR is failed with XR gem. I believe there is a commit which fixes it. You will need merge the change from development branch.

Thanks for the info @VickyAtAZ, let me know when you've done this @michalpelka and I can kick off another build 👍

michalpelka commented:
@VickyAtAZ thanks for suggestion.
@hultonha just rebased on top of #103, which was rebased on top of development.

hultonha commented:
> @VickyAtAZ thanks for suggestion. @hultonha just rebased on top of #103, which was rebased on top of development.

Build started - https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-117/2/

michalpelka commented:
Rebased manually, and tested locally.

spham-amzn commented on Gems/ROS2/Code/Source/Camera/CameraSensor.cpp:
```diff
@@ -116,11 +117,11 @@ namespace ROS2
 
-        const AZStd::string pipelineName = m_cameraSensorDescription.m_cameraName + "Pipeline" + GetPipelineTypeName();
-
+        m_pipelineName = AZStd::string::format("%sPipeline%s%s",m_cameraSensorDescription.m_cameraName.c_str(), GetPipelineTypeName().c_str(),
```
nit: Is it possible to move this to the constructor initializer list? (if `GetPipelineTypeName()` doesnt change)

spham-amzn commented on Gems/ROS2/Code/Source/Camera/CameraSensor.h:
```diff
@@ -57,28 +59,29 @@ namespace ROS2
-        void RequestFrame(
-            const AZ::Transform& cameraPose, AZStd::function<void(const AZ::RPI::AttachmentReadback::ReadbackResult& result)> callback);
+        void RequestMessagePublication(
+            std::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::Image>> publisher,
```
nit:
```suggestion
            AZStd::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::Image>> publisher,
```

spham-amzn commented on Gems/ROS2/Code/Source/Camera/CameraSensor.h:
```diff
@@ -87,6 +90,27 @@ namespace ROS2
+
+        //! Request a frame from the rendering pipeline
+        //! @param cameraPose - current camera pose from which the rendering should take place
+        //! @param callback - callback function object that will be called when capture is ready
+        //!                   it's argument is readback structure containing, among other thins, captured image
```
```suggestion
        //! @param callback - callback function object that will be called when capture is ready.
        //!                   It's argument is readback structure containing, among other things, a captured image
```

spham-amzn commented on Gems/ROS2/Code/Source/Camera/CameraSensor.h:
```diff
@@ -87,6 +90,27 @@ namespace ROS2
+            const AZ::Transform& cameraPose, AZStd::function<void(const AZ::RPI::AttachmentReadback::ReadbackResult& result)> callback);
+
+        //! Request an additional frame from the rendering pipeline
+        //! @param cameraPose - current camera pose from which the rendering should take place
```
Remove this line, there is no `cameraPose` argument in this function

spham-amzn commented on Gems/ROS2/Code/Source/Camera/ROS2CameraSensorComponent.cpp:
```diff
@@ -100,9 +115,9 @@ namespace ROS2
+        if (!m_imagePublishers.empty() && m_cameraSensor)
         {
-            const auto& camera_descritpion = m_cameraSensorsWithPublihsers.front().second->GetCameraSensorDescription();
+            const auto& camera_descritpion = m_cameraSensor->GetCameraSensorDescription();
```
```suggestion
            const auto& camera_description = m_cameraSensor->GetCameraSensorDescription();
```

lumberyard-employee-dm replied:
Also the variable naming should be lowerCamelCase

spham-amzn reviewed (approved):
Looks good, just minor suggestions

michalpelka commented:
Thank you @spham-amzn for your review. Applied your recommendation where possible and explained where it was not feasible. 
I also rebased manually and adjusted PR's code to recent changes in frame capture API.

adamdbrw commented:
@hultonha it seems that the build failed due to the Asset Processor failing with the AtomContent Gem (`Gems/AtomContent/TestData`). It seems to be unrelated. What could we do to make sure the AR goes through?

hultonha commented:
> @hultonha it seems that the build failed due to the Asset Processor failing with the AtomContent Gem (`Gems/AtomContent/TestData`). It seems to be unrelated. What could we do to make sure the AR goes through?

Hmm so the best thing I can suggest is to report the issue for the Atom team to follow-up about (see this template - https://github.com/o3de/o3de/issues/new?assignees=&labels=needs-triage%2Ckind%2Fbug%2Ckind%2Fautomation&template=ar_bug_report.md&title=AR+Bug+Report) and then merge development into your branch (or rebase) and I can then kick off another build and hope it doesn't get hit again. Can you let me know when that's done and I can start another build. Thanks!

michalpelka commented:
@hultonha rebased just now.

hultonha commented:
Great to hear @michalpelka, just building now 👍 https://jenkins.build.o3de.org/view/Projects/job/o3de-extras/view/change-requests/job/PR-117/5/

michalpelka commented:
Unfortunately, there was a failed build test in AR, which seemed unrelated.
I rebased, could yo try to rebuild @hultonha ?
